twister: coverage: Data collection and reporting per-test instance

golowanow opened the pull request:
Twister code coverage data collection and reporting improvements:

With this change, the coverage data (GCOV dump) is extracted from the test log files after each of the test instance parallel execution, instead of post processing all the logs at the end of the Twister run.

1) The new `--coverage-per-instance` mode extend Twister code coverage operations to ~~collect and~~ report statistics on each test instance execution individually in addition to the default reporting mode of data aggregation for all the test instances in current scope of the Twister run.
The split mode allows to identify precisely what amount of code coverage each test case provides and to analyze its contribution to the overall test plan's coverage. Each build directory will have its own coverage report and data files, so the overall disk space and the total execution time increase.
In this mode the aggregate code coverage report for 'gcovr' is merged from per-instances' coverage.json reports instead of processing .gcda data files again.

Another new `--disable-coverage-aggregation` option allows to execute only the `--coverage-per-instance` mode if the summary statistics are not needed.

~~2) New Twister option `--coverage-sections` allows to select what coverage data to include into the `twister_coverage.json` file for its further analysis together with the test meta-data and test results.
The choices are to select either the coverage summary, or the detailed branch coverage, ztest coverage, or all the coverage data collected. Also the coverage run 'status' and coverage 'tool' values are added with any non-default choice. The 'environment' top object has additional 'gcov_tool' property to keep the command line argument selection.
Depending on `--coverage-split` and `--disable-coverage-aggregation` options, the coverage data is attached either to its test suite object or/and at the report's top level as the aggregated summary of the test plan execution scope.
Currently this mode is fully supported for the default gcovr reporting tool only, as it is based on the GCOVR json reports. For lcov only coverage run status is reported.~~

The new reporting modes are disabled by default for compatibility.

3) Group Twister code coverage command line arguments for better usability. Add more details in descriptions.

partially resolves #67058

golowanow committed:
twister: coverage: Data collection and reporting per-test instance

With this change, the coverage data (GCOV dump) is extracted from the
test log files after each of the test instance parallel execution,
instead of post processing all the logs at the end of the Twister run.

The new `--coverage-per-instance` mode extends Twister coverage operations
to report coverage statistics on each test instance execution individually
in addition to the default reporting mode which aggregates data to one
report with all the test instances in the current scope of the Twister run.
The split mode allows to identify precisely what amount of code coverage
each test suite provides and to analyze its contribution to the overall
test plan's coverage. Each test configuration's output directory will have
its own coverage report and data files, so the overall disk space and
the total execution time increase.

Another new `--disable-coverage-aggregation` option allows to execute
only the `--coverage-per-instance` mode when the aggregate coverage
report for the whole Twister run scope is not needed.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
for example, with `./scripts/twister -vv -j 1 -p unit_testing  --coverage --coverage-tool gcovr -T tests/unit/list --coverage-split --coverage-report summary`:

twister.json
````
{
....
    "testsuites":[
        {
            "name":"tests/unit/list/utilities.lists",
            "arch":"unit",
            "platform":"unit_testing",
   ....
            "testcases":[
                {
                    "identifier":"utilities.lists.slist",
                    "execution_time":"0.00",
                    "status":"passed"
                },
                {
                    "identifier":"utilities.lists.dlist",
                    "execution_time":"0.00",
                    "status":"passed"
                },
....
            ],
            "coverage":{
                "status":true,
                "summary":{
                    "branch_covered":158,
                    "branch_percent":52.3,
                    "branch_total":302,
                    "files":[
                        {
                            "branch_covered":24,
                            "branch_percent":100.0,
                            "branch_total":24,
                            "filename":"include/zephyr/sys/dlist.h",
                            "function_covered":21,
                            "function_percent":100.0,
                            "function_total":21,
                            "line_covered":82,
                            "line_percent":100.0,
                            "line_total":82
                        },
```
                    ],
                    "function_covered":91,
                    "function_percent":80.5,
                    "function_total":113,
                    "gcovr/summary_format_version":"0.5",
                    "line_covered":394,
                    "line_percent":61.0,
                    "line_total":646,
                    "root":"../../../../../.."
                }
            }
        }
    ],
    "coverage":{
        "status":true,
        "summary":{
            "branch_covered":158,
            "branch_percent":52.3,
            "branch_total":302,
            "files":[
                {
....
                }
            ],
            "function_covered":91,
            "function_percent":80.5,
            "function_total":113,
            "gcovr/summary_format_version":"0.5",
            "line_covered":394,
            "line_percent":61.0,
            "line_total":646,
            "root":".."
        }
    }
}
```

golowanow commented:
@gchwier, @gopiotr, @hakehuang, @nashif, @PerMac up to your attention

hakehuang commented:
@golowanow , maybe change the PR title to enable json report with coverage for each file?

golowanow commented:
> @golowanow , maybe change the PR title to enable json report with coverage for each file?

That title will be misleading as gcovr coverage report already has details for each file, and now it is included into `twister.json`.
The  example snippet above is to show the gcovr 'summary' mode results embedded.

hakehuang commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -313,6 +313,25 @@ def add_parse_arguments(parser = None):
                              "Default to html. "
                              "Valid options are html, xml, csv, txt, coveralls, sonarqube, lcov.")
 
+    parser.add_argument("--disable-coverage-aggregation", action="store_true", default=False,
```
would this be better to group the new added options to a cov_opt = parser.add_argument_group()

golowanow replied:
done

hakehuang reviewed (dismissed):
LGTM, with some code suggestion on group coverage options. beside the conflict need to be resolved

golowanow committed:
twister: coverage: Group command line arguments

Group Twister code coverage command line arguments for better usability.
Add more details in descriptions.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
rebased to resolve conflicts with #66487 and with some minor improvements: 
* command line arguments re-arranged; 
* additional properties added to the json (to keep which gcov-tool and gcovr/lcov was used).

golowanow commented:
rebased resolving merge conflict with #66898

golowanow commented:
rebased resolving merge conflict with #66905

golowanow commented:
> @golowanow do you plan to return to this PR?

yes, I do. I'm ready to respond to review comments on this proposal.

PerMac commented:
@nordic-piks Maybe you'd like to have a look at this?

golowanow commented:
rebased to resolve merge conflicts (with #67129), some minor corrections and adjustments to twister tests.

@hakehuang, please refresh your +1

golowanow commented:
@gchwier, @gopiotr, @nashif, @PerMac, any objections or comments on this PR ? 
some more details are also in the RFC #67058

katgiadla commented:
> @gchwier, @gopiotr, @nashif, @PerMac, any objections or comments on this PR ? some more details are also in the RFC #67058

I want to check it before PR merge. Please give me about 2 days.

golowanow commented:
> I want to check it before PR merge. Please give me about ~2~3 days.

@katgiadla, did you had time to check ?

katgiadla commented:
> > I want to check it before PR merge. Please give me about ~2~3 days.
> 
> @katgiadla, did you have time to check ?

No, I hadn't. I'll check it at this week.

golowanow commented:
@katgiadla, do you have any feedback to share ?

kartben commented:
This needs rebasing but not sure what's the current status of this PR?

cfriedt commented:
@golowanow - friendly ping to please rebase

golowanow commented:
1) rebased;
2) save coverage to separate `twister_coverage.json` instead of adding to `twister.json` - the same way as for memory footprints.

^^ @hakehuang, @nashif, @PerMac

nashif reviewed (changes requested):
per test coverage analysis and processing data using the queue is fine. We should not be capturing this data however just because we can, there are tools that process this data and display it already, so lets use those tools. 

#84106 shows how this can be done with available analysis and visualization tools.

golowanow commented:
> per test coverage analysis and processing data using the queue is fine. We should not be capturing this data however just because we can, there are tools that process this data and display it already, so lets use those tools.
> 
> #84106 shows how this can be done with available analysis and visualization tools.

@nashifl, It is totally not clear what kind of changes you request with your -1 at this PR now. 
Did you read the preamble or the related RFC ?

This change allows to use the same old-known html reports (grovr or lcov) as well as to transfer the coverage data for analysis by more powerful tools, not mention machine learning.

nashif commented:
> @nashif, It is totally not clear what kind of changes you request with your -1 at this PR now.

I said:

> We should not be capturing this data however just because we can, there are tools that process this data and display it already, so lets use those tools.

Changes I request: Twister shouild not capture or deal with exporting such data that we already have tools for and that is not directly related to test results.

golowanow commented:
> Changes I request: Twister shouild not capture or deal with exporting such data that we already have tools for and that is not directly related to test results.

`twister_coverage.json` part removed from this PR

nashif commented:
few things:
1. running with --coverage-split now is VERY verbose, need to get rid of all those log messages....
```
INFO    - Using gcov tool: gcov  13  15%  built (not run):    0, filtered:    7, failed:    0, error:    0
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
ERROR   - GCOVR failed with 64
INFO    - Using gcov tool: gcov  13  23%  built (not run):    0, filtered:    7, failed:    0, error:    0
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
INFO    - Using gcov tool: gcov
INFO    - Generating coverage files...
```
2. I think it makes sense so enable this functionality by default and have coverage info generated as part of the run queue and do the collection at the end, basically keep current behaviour of generating reports as we have it now, but benefit from parallel execution and generation.
3. whether we generate a report (html) for each instance is managed via an option
4. Where supported (lcov), integrate test data, similar to what is done in #84106. Probably that can be done as a followup and a reuse of #84106 if this one get merged first. @jeremybettis what do you think?

golowanow commented:
> 1. running with --coverage-split now is VERY verbose, need to get rid of all those log messages....

removed or changed to debug.

> 2. I think it makes sense so enable this functionality by default and have coverage info generated as part of the run queue and do the collection at the end, basically keep current behaviour of generating reports as we have it now, but benefit from parallel execution and generation.
> 3. whether we generate a report (html) for each instance is managed via an option

changed to do it this way.

> btw, I see GCOVR failed with 64 all over the place now, need to see what the issue is, 

I haven't got these with my tests using gcovr 8.2 and gcov from SDK 0.17.0.
What was the twister command line when you got these errors ?

nashif commented:
> What was the twister command line when you got these errors ?

 ./scripts/twister -p native_sim -T tests/kernel/threads/ --coverage --coverage-split -v

nashif commented:
with current PR, I get:

```
./scripts/twister -p native_sim -T tests/kernel/threads/ --coverage --coverage-split -v
Renaming output directory to /home/anashif/i9/zephyrproject/zephyr/twister-out.1
INFO    - Using Ninja..
INFO    - Zephyr version: v4.0.0-3521-gdd1ef76e6fef
INFO    - Using 'llvm' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/anashif/i9/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 22
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    -  1/13 native_sim/native         tests/kernel/threads/thread_error_case/kernel.threads.error.case FILTERED (runtime filter)
INFO    -  2/13 native_sim/native         tests/kernel/threads/tls/kernel.threads.tls        FILTERED (runtime filter)
ERROR   - General exception: argument should be a str or an os.PathLike object where __fspath__ returns a str, not 'NoneType'
ERROR   - General exception: argument should be a str or an os.PathLike object where __fspath__ returns a str, not 'NoneType'
ERROR   - General exception: argument should be a str or an os.PathLike object where __fspath__ returns a str, not 'NoneType'
ERROR   - Process 191985 failed, aborting execution
```

golowanow commented:
> ./scripts/twister -p native_sim -T tests/kernel/threads/ --coverage --coverage-split -v

It runs for me.
Is it because of your `INFO    - Using 'llvm' toolchain.` ?

```
$ ./scripts/twister -p native_sim -T tests/kernel/threads/ --coverage --coverage-split -v
INFO    - Using Ninja..
INFO    - Zephyr version: v4.0.0-3521-gdd1ef76e6fef
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /foo/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 12
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    -  1/13 native_sim/native         tests/kernel/threads/tls/kernel.threads.tls        FILTERED (runtime filter)
INFO    -  2/13 native_sim/native         tests/kernel/threads/thread_error_case/kernel.threads.error.case FILTERED (runtime filter)
INFO    - HTML report generated: /foo/zephyrproject/zephyr/twister-out/native_sim_native/host/tests/kernel/threads/thread_init/kernel.threads.init/coverage/index.html
INFO    -  3/13 native_sim/native         tests/kernel/threads/thread_init/kernel.threads.init PASSED (native 0.168s <host>)
INFO    - HTML report generated: /foo/zephyrproject/zephyr/twister-out/native_sim_native/host/tests/kernel/threads/thread_stack/kernel.threads.thread_stack/coverage/index.html
...
INFO    - HTML report generated: /foo/zephyrproject/zephyr/twister-out/native_sim_native/host/tests/kernel/threads/thread_apis/kernel.threads.apis/coverage/index.html
INFO    - 13/13 native_sim/native         tests/kernel/threads/thread_apis/kernel.threads.apis PASSED (native 0.156s <host>)

INFO    - 18 test scenarios (18 configurations) selected, 7 configurations filtered (5 by static filter, 2 at runtime).
INFO    - 11 of 11 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 62.99 seconds.
INFO    - 54 of 54 executed test cases passed (100.00%) on 1 out of total 911 platforms (0.11%).
INFO    - 26 selected test cases not executed: 26 skipped.
INFO    - 11 test configurations executed on platforms, 0 test configurations were only built.
INFO    - HTML report generated: /foo/zephyrproject/zephyr/twister-out/coverage/index.html
INFO    - Saving reports...
INFO    - Writing JSON report /foo/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /foo/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /foo/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

nashif commented:
> with current PR, I get:

ignore(! maybe not), I had LLVM set as a variant...

going back to zephyr variant, I get:

```
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    -  1/13 native_sim/native         tests/kernel/threads/thread_error_case/kernel.threads.error.case FILTERED (runtime filter)
INFO    -  2/13 native_sim/native         tests/kernel/threads/tls/kernel.threads.tls        FILTERED (runtime filter)
ERROR   - GCOVR failed with 64
INFO    -  3/13 native_sim/native         tests/kernel/threads/dynamic_thread_stack/kernel.threads.dynamic_thread.stack.pool.no_alloc.user PASSED (native 0.049s <host>)
ERROR   - GCOVR failed with 64
INFO    -  4/13 native_sim/native         tests/kernel/threads/dynamic_thread_stack/kernel.threads.dynamic_thread.stack.no_pool.alloc.user PASSED (native 0.209s <host>)
ERROR   - GCOVR failed with 64
ERROR   - GCOVR failed with 64
```

gcovr: 8.2

nashif commented:
> It runs for me.

hmm, strange.

btw, still having too verbose messaging, i.e. 

```
HTML report generated: ...
```

> > 3\. I think it makes sense so enable this functionality by default and have coverage info generated as part of the run queue and do the collection at the end, basically keep current behaviour of generating reports as we have it now, but benefit from parallel execution and generation.
> > 4\. whether we generate a report (html) for each instance is managed via an option
> 
> changed to do it this way.

this seems to be the same old behaviour as before, i.e.. running with `--coverage` only should imply what `--coverage-split` now does.

golowanow commented:
> this seems to be the same old behaviour as before, i.e.. running with `--coverage` only should imply what `--coverage-split` now does.

right, corrected.

> btw, still having too verbose messaging, i.e.
> 
> ```
> HTML report generated: ...
> ```

I left it the same INFO as other logs for reports:
```
INFO - LCOV report generated
INFO - HTML report generated
INFO - Saving reports...   
INFO - Writing JSON report 
INFO - Writing xunit report
INFO - Writing xunit report
```

golowanow committed:
twister: coverage: Merge gcovr reports

In `--coverage-per-instance` mode for 'gcovr' merge individual
coverage.json reports into the aggregate code coverage report
instead of processing .gcda data files again.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
> .. for gcovr, is to merge coverage.json at instances to the summary report to speed it up.

added to this PR as well.

nashif commented on scripts/pylib/twister/twisterlib/coverage.py:
```diff
@@ -289,12 +304,15 @@ def _generate(self, outdir, coveragelog):
-        return self.run_command(cmd, coveragelog)
+        ret = self.run_command(cmd, coveragelog)
+
+        # TODO: Add support to include LCOV .info files into JSON data.
```
what is that? What json data?

golowanow replied:
changed to remind to call lcov for summary coverage report similar to what gcovr makes.

nashif reviewed (commented):
change `--coverage-split` to `--coverage-per-instance` or `--coverage-per-testsuite`, after all we are not splitting anything here.

@jeremybettis are you ok with this change and with a possible followup incorporating your enhancements in the other PR?

jeremybettis commented:
> change `--coverage-split` to `--coverage-per-instance` or `--coverage-per-testsuite`, after all we are not splitting anything here.
> 
> @jeremybettis are you ok with this change and with a possible followup incorporating your enhancements in the other PR?

Yes, I don't mind redoing my change on top of this.

golowanow commented:
^^ @hakehuang. @PerMac - could you please look at this (again) ?
